src/appleseed/foundation/utility/zipper.cpp

```diff
@@ -0,0 +1,345 @@
+#include "foundation/utility/string.h"
+
+// Boost headers.
+#include "boost/filesystem.hpp"
+
+// Standard headers.
+#include <ctime>
+#include <vector>
+#include <fstream>
+#include <set>
+
+using namespace std;
+namespace bf = boost::filesystem;
+
+namespace foundation
+{
+
+    ZipException::ZipException(const char* what)
+        : Exception(what)
+    {
+    }
+
+    ZipException::ZipException(const char* what, const int err)
+    {
+        string string_what = what + to_string(err);
+        set_what(string_what.c_str());
+    }
+
+    bool is_zip_entry_directory(const string& dirname)
+    {
+        // used own implementation of is_zip_entry_directory instead of boost implementation
+        // because this directory is not in filesystem, but in zipfile
+        return dirname[dirname.size() - 1] == '/';
+    }
+
+    void open_current_file(unzFile& zip_file)
+    {
+        const int err = unzOpenCurrentFile(zip_file);
+        if (err != UNZ_OK)
+            throw ZipException("Can't open file inside zip: ", err);
+    }
+
+    int read_chunk(unzFile& zip_file, char* buffer, const int chunk_size)
+    {
+        const int err = unzReadCurrentFile(zip_file, buffer, chunk_size);
+
+        if (err == UNZ_ERRNO)
+            throw ZipException("IO error while reading from zip");
+        if (err < 0)
+            throw ZipException("zLib error while decompressing file: ", err);
+
+        return err;
+    }
+
+    void unzip_close_current_file(unzFile& zip_file)
+    {
+        const int err = unzCloseCurrentFile(zip_file);
+
+        if (err == UNZ_CRCERROR)
+            throw ZipException("CRC32 is not good");
+    }
+
+    string read_filename(unzFile& zip_file)
+    {
+        unz_file_info zip_file_info;
+        unzGetCurrentFileInfo(zip_file, &zip_file_info, 0, 0, 0, 0, 0, 0);
+
+        vector<char> filename(zip_file_info.size_filename + 1);
+        unzGetCurrentFileInfo(
+            zip_file,
+            &zip_file_info,
+            &filename[0],
+            static_cast<uLong>(filename.size()),
+            0, 0,
+            0, 0);
+        filename[filename.size() - 1] = '\0';
+
+        return string(&filename[0]);
+    }
+
+    string get_filepath(unzFile& zip_file, const string& unzipped_dir)
+    {
+        string filename = read_filename(zip_file);
+        return (bf::path(unzipped_dir) / bf::path(filename)).string();
+    }
+
+    void extract_current_file(unzFile& zip_file, const string& unzipped_dir)
+    {
+        const string filepath = get_filepath(zip_file, unzipped_dir);
+
+        if (is_zip_entry_directory(filepath))
+        {
+            bf::create_directories(filepath);
+            return;
+        }
+        else open_current_file(zip_file);
+
+        fstream out(filepath.c_str(), ios_base::out | ios_base::binary);
+        if (out.fail())
+            throw ZipException(("Can't open file " + filepath).c_str());
+
+        const size_t BUFFER_SIZE = 4096;
```

Comment:
Nope, allocating one more stack variable is free, regardless of where the variable is allocated in the function.

Stack allocation is performed by a single pointer subtraction (of the maximum amount of stack memory required by the function) at the entry of the function.

A few additional details here: http://stackoverflow.com/a/80113/393756